File: 0152-maximum-product-subarray/0152-maximum-product-subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int n = nums.length;
        double p = 1, s = 1;
        double max = Integer.MIN_VALUE; 
        for (int i = 0; i < n; i++) {
            p *= nums[i];
            s *= nums[n - i - 1];
            max = Math.max(max, Math.max(p, s));
            if (p == 0) p = 1; 
            if (s == 0) s = 1; 
        }
        return (int)max;
    }
}

File: 0152-maximum-product-subarray/0152-maximum-product-subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int prod = 1 , maxi = INT_MIN;
        for(int i : nums) {
            prod *= i;
            maxi = max(maxi , prod);
            if(prod == 0) prod = 1; 
        }
        prod = 1;
        for(int i = nums.size() - 1 ; i >= 0 ; i --) {
            prod *= nums[i];
            maxi = max(maxi , prod);
            if(prod == 0) prod = 1;
        }
        return maxi;
    }
};